Web Extensions: Clean up Mock Node and Tests for Get and Create for the Bookmarks API

akutira-umich opened the pull request:
#### 60367f4ebf3c8581a95ce18db1a223da82b38d62
<pre>
Web Extensions: Clean up Mock Node and Tests for Get and Create for the Bookmarks API
and add getRecent
<a href="https://rdar.apple.com/157304875">rdar://157304875</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=296813">https://bugs.webkit.org/show_bug.cgi?id=296813</a>

Reviewed by Brian Weinstein.

Moved all mock node implementation that was present in Create and getRecent in the Bookmarks API&apos;s WebProcess to the TestWebKitAPI space. Changed getTree to return a vector of its top level nodes instead of the root to be in line with MDN documentation. Implemented getRecent in WebProcess and UIProcess which included adding the dateAdded parameter to the BookmarkTreeNode list of parameters.

* Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::createParametersFromProtocolObject):
(WebKit::flattenAndConvertAllBookmarks):
(WebKit::WebExtensionContext::bookmarksGetTree):
(WebKit::WebExtensionContext::bookmarksGetRecent):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::toAPI):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
(WebKit::WebExtensionAPIBookmarks::getTree):
(WebKit::toWebAPI): Deleted.
(WebKit::WebExtensionAPIBookmarks::initializeMockBookmarksInternal): Deleted.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
(): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(-[_MockBookmarkNode dateAddedForWebExtensionContext:]):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureCreateBookmarkDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIgetRecent)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreate)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetTree)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateAndGetTree)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetSubTreeChildren)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMockNodeWithgetRecent)): Deleted.
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateParse)): Deleted.
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, CreateAndGetTree)): Deleted.
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, GetSubTree)): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::createParametersFromProtocolObject):
(WebKit::WebExtensionContext::bookmarksGetRecent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(-[_MockBookmarkNode bookmarkTypeForWebExtensionContext:]):
(-[_MockBookmarkNode childrenForWebExtensionContext:]):
(-[_MockBookmarkNode dateAddedForWebExtensionContext:]):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureCreateBookmarkDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetRecent)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIgetRecent)): Deleted.

Canonical link: <a href="https://commits.webkit.org/298290@main">https://commits.webkit.org/298290@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bfbc6faa92b43c62b68dd100e60e4b4fca40c8b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34626 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25100 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121032 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65583 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116780 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35262 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43185 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87311 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42162 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; layout-tests (exception)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28095 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103161 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67704 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27271 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21282 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64689 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97477 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124225 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41885 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31301 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96113 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42256 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99351 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95898 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24428 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41096 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18930 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37931 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41759 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47288 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41312 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44628 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43055 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -46,6 +46,12 @@
     if (NSArray *children = [bookmark childrenForWebExtensionContext:context])
         parameters.children = createParametersFromProtocolObjects(children, context);
 
+    NSDate *dateAddedNSDate = [bookmark dateAddedForWebExtensionContext:context];
```
Nit: This could just be `NSDate *dateAdded`.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -337,7 +362,55 @@
+    if (![controllerDelegate respondsToSelector:@selector(_webExtensionController:bookmarksForExtensionContext:completionHandler:)]) {
+        completionHandler(toWebExtensionError(apiName, nullString(), @"it is not implemented"));
+        return;
+    }
```
Nit: You could add a blank line here.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -337,7 +362,55 @@
+
+        std::sort(allBookmarksParameters.begin(), allBookmarksParameters.end(), [](const WebExtensionBookmarksParameters& a, const WebExtensionBookmarksParameters& b) {
+            return a.dateAdded > b.dateAdded;
+        });
```
Did you test this for performance at all? For users with massive bookmark sets, it could be slow to give all the bookmarks to WebKit, convert them to `WebExtensionBookmarksParameters`, sort them, and then drop most of them on the floor.

akutira-umich replied:
My performance testing only covered up to 600 bookmarks. I didn't have any issues with that. Although, I'm not sure of an alternative then just looking at the full list and grabbing only the `numberofItem`'s recent bookmarks

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -80,6 +64,8 @@
         dictionary[urlKey] = node.url->createNSString().get();
         dictionary[typeKey] = @"bookmark";
+        double seconds = node.dateAdded.secondsSinceEpoch().value();
+        dictionary[dateAddedKey] = @(seconds * 1000.0);
```
Do folders not have a date added?

akutira-umich replied:
No folders do not have the data added property

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -130,120 +93,32 @@
-
-    double msSinceEpoch = dynamic_objc_cast<NSNumber>(bookmark[dateAddedKey]).doubleValue;
-    newNode->dateAdded = WTF::WallTime::fromRawSeconds(msSinceEpoch / 1000.0);
+    const std::optional<String>& parentmockId = bookmark[parentIdKey];
```
Nit: `parentMockId` or `parentMockID`

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -130,120 +93,32 @@
-    const std::optional<String>& title = newNode->title;
-    const std::optional<String>& url = newNode->url;
-
-    std::optional<uint64_t> index = (newNode->index);
```
Lot of red here which is nice to see!

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -288,25 +163,26 @@
-    NSMutableArray *resultArray = [NSMutableArray arrayWithCapacity:static_cast<NSUInteger>(numberOfItems)];
+            for (const auto& bookmarkParams : recentBookmarksParams) {
+                NSDictionary* bookmarkDictionary = toAPI(bookmarkParams);
+                [resultArray addObject:bookmarkDictionary];
```
Super-nit: I think this could just be `[resultArray addObject:toAPI(bookmarkParams)];`

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -85,6 +85,7 @@ - (_WKWebExtensionBookmarkType)bookmarkTypeForWebExtensionContext:(WKWebExtensio
     return _WKWebExtensionBookmarkTypeFolder;
 }
+
 - (NSArray<id<_WKWebExtensionBookmark>> *)childrenForWebExtensionContext:(WKWebExtensionContext *)context {
```
While you're here, can you move the `{` down to the next line?

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -95,6 +96,18 @@ - (_WKWebExtensionBookmarkType)bookmarkTypeForWebExtensionContext:(WKWebExtensio
 
+- (NSDate *)dateAddedForWebExtensionContext:(WKWebExtensionContext *)context
+{
+    id dateValue = _dictionary[@"dateAdded"];
```
`objectForKey<NSDate>`?

b-weinstein replied:
Sorry, `NSNumber` - since it looks like this vends a number.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -254,6 +267,9 @@ void configureCreateBookmarkDelegate(TestWebExtensionManager *manager)
             newBookmarkData[@"title"] = title;
             if (url)
                 newBookmarkData[@"url"] = url;
+            double dateAddedInMilliseconds = [[NSDate date] timeIntervalSince1970] * 1000.0;
```
Nit: Does `NSDate.date.timeIntervalSince1970` compile and work?

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -381,16 +397,17 @@ void configureGetBookmarksDelegate(TestWebExtensionManager *manager)
 }
 
-TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMockNodeWithgetRecent)
+TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIgetRecent)
```
Nit: `BookmarksAPIGetRecent`

b-weinstein reviewed (changes requested):
A few small things but this is very close!

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h:
```diff
@@ -74,6 +74,13 @@ WK_SWIFT_UI_ACTOR
+ @param context The context in which the web extension is running.
+ @return The date the bookmark was added. Should be `nil` for folders or separators.
+ */
+- (nullable NSDate *)dateAddedForWebExtensionContext:(WKWebExtensionContext * _Nonnull)context NS_SWIFT_NAME(dateAdded(for:));
```
Is there a reason we don't have `dateAdded` for folders? The node documentation (https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks/BookmarkTreeNode) seems to indicate that folders should have a `dateAdded` as well.

akutira-umich replied:
Couple reasons why this is the case. DateAdded is only really used for `bookmarks.getRecent()` function which (according to the MDN docs) "specified number of the most recently added bookmarks". Another reason is that the `dateAdded` parameter doesn't exist in the `SafariWebBookmarkList` implementation. `dataAdded` is only available in `SafariWebBookmark.h`. Also Ari sent me this convo https://a1391192.slack.com/archives/C07STK4J0TZ/p1749145814775789 (I don't have access to it) that explains why we won't be able to use `dateAdded` for folders.

b-weinstein replied:
Going back to the documentation again (https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks/getRecent) - I don't see anything that indicates that this is only leaf bookmarks.

Is this behavior something that you tested in other browsers? If so - that should be mentioned in the commit message somewhere, because it's unclear where this came from, as it doesn't appear to be backed by the documentation.

akutira-umich committed:
Web Extensions: Clean up Mock Node and Tests for Get and Create for the Bookmarks API
and add getRecent
rdar://157304875
https://bugs.webkit.org/show_bug.cgi?id=296813

Reviewed by Brian Weinstein.

Moved all mock node implementation that was present in Create and getRecent in the Bookmarks API's WebProcess to the TestWebKitAPI space. Changed getTree to return a vector of its top level nodes instead of the root to be in line with MDN documentation. Implemented getRecent in WebProcess and UIProcess which included adding the dateAdded parameter to the BookmarkTreeNode list of parameters.

* Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::createParametersFromProtocolObject):
(WebKit::flattenAndConvertAllBookmarks):
(WebKit::WebExtensionContext::bookmarksGetTree):
(WebKit::WebExtensionContext::bookmarksGetRecent):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::toAPI):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
(WebKit::WebExtensionAPIBookmarks::getTree):
(WebKit::toWebAPI): Deleted.
(WebKit::WebExtensionAPIBookmarks::initializeMockBookmarksInternal): Deleted.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
(): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(-[_MockBookmarkNode dateAddedForWebExtensionContext:]):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureCreateBookmarkDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIgetRecent)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreate)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetTree)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateAndGetTree)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetSubTreeChildren)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMockNodeWithgetRecent)): Deleted.
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateParse)): Deleted.
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, CreateAndGetTree)): Deleted.
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, GetSubTree)): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::createParametersFromProtocolObject):
(WebKit::WebExtensionContext::bookmarksGetRecent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(-[_MockBookmarkNode bookmarkTypeForWebExtensionContext:]):
(-[_MockBookmarkNode childrenForWebExtensionContext:]):
(-[_MockBookmarkNode dateAddedForWebExtensionContext:]):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureCreateBookmarkDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetRecent)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIgetRecent)): Deleted.

Canonical link: https://commits.webkit.org/298290@main
